### DISCUSSION TITLE
Discussion : Name constructor with keyword instead of name

### DISCUSSION THREAD

**Discussion by ebfortin** [dotnet/csharplang#1928] [Language Ideas] [answered]
I find it tedious to repeat the name of a class each time I declare a constructor. IDE helps but still, I see no added value to just repeat the class name. I propose the addition of "ctor" keyword to be used in place of the constructor name. Could also use "dtor" for finalizer.

Even IL understands it. It already use that syntax.

Anyone has an opinion on this?

**Comment by theunrepentantgeek** [comment]
@ebfortin Can you please clarify? Are you talking about use of the type name when _constructing_ an instance, or when _declaring_ the constructor as a part of the definition?

I thought it was the later, but the comment from @jnm2 has me questioning whether I read your proposal correctly.

**Comment by CyrusNajmabadi** [comment, marked as answer]
It seems highly unlikely C# would change here.  Having two ways to do something, where the new way provides only a tiny amount of marginal benefit, would likely not be desirable.

**Reply by AJLProjects** [reply]
"two ways to do something" - What about the var keyword, that is analogous in a way. "Tiny amount of marginal benefit", how many classes are there in existance? A lot. This might be functionally the same, but it is significant.

**Reply by HaloFour** [reply]
The `var` keyword was added primarily to support anonymous types which are nameless and thus you cannot declare a variable of a type of them directly.  The ability to use them for inferred variables in general was just an added bonus.

**Reply by AJLProjects** [reply]
>  just an added bonus.

Added bonus but still a competing convention none the less.

In some circumstancez competing conventions is bad, but there is no drawback to having constructor along with the normal way of constructing. Since everyone knows what constructor is. (I hope)

**Reply by Rekkonnect** [reply]
I've stumbled upon this problem quite a few times, copying some constructors over to another place, where the only thing I have to change is the name of the constructed type. Plus, it's quite tedious seeing the same name over and over, neither is it intuitive to differentiate it between a ctor and a method with return type being the declared one.

However, I'm more in favor of using `new` for construction (nothing for the destructor I could come up with), and so:

Proposed syntax:
```csharp
public new(int value, string name)
{
}
// instead of
public ValueName(int value, string name)
{
}
```

This would also apply to the static ctor, which would allow the following syntax:
```csharp
static new()
{
}
```

In other words, benefits of permitting ctor declaration with the syntax `new(...)`:
- easier to copy-paste over
- much clearer about whether it's a method or a ctor
- helpful for types with large names
- matches the target-typed ctor syntax introduced in C# 9.0
- better applies the DRY principle

The drawbacks are:
- `new` cannot be used as a modifier in ctors like it can be used for hiding base members (if the language ever, mistakenly, chooses to support this)

## Considerations

For the `static` constructor, using `new` semantically does not make much sense, since it's not an instance constructor. In fact, Java prefers to call it a static initialization block, and its syntax is `static { }`. It could be adopted in a similar fashion, with the following options:

```csharp
static
{
}
```
```csharp
static()
{
}
```

The option to keep the parentheses is more consistent with the language, and more clearly indicates a method. However, the static initialization block (static constructor), cannot accept any arguments.

On the other hand, the simple `static { }` syntax reserves the ability to add a potential future feature where multiple members are declared in a block with specific modifiers, i.e.
```csharp
public static
{
    void M();
    void M1(int arg);
}
```

**Reply by NinjaCross** [reply]
IMHO, in light of the recent evolutions in the C# language, this topic ought to be re-opened/re-examined.
As far as I can see, the global evolution of the language lead to a "concise at any cost" approach, which I personally appreciate very much.
Maybe in 2018, when the proposal has been initially made, this topic was somehow of secondary importance.
Now I believe its' more relevant/actual.
By mistake yesterday I made the same proposal (I couldn't find this thread, so I created a new one  #5361 ) providing also a realistic use case.
For the sake of simplicity, I closed the discussion  #5361, and I report here below what I think is a good reasoning framewok for this topic.

---


Personally, I'm all about conciseness, and I think that classes with (very) long names should generally be avoided, 
I have to admit anyway that sometimes a long name is much more usefull than a shorter one, and helps to understand the nature and intention of the object just reading at its name.

I will try to explain why I think we should introduce the keyword "constructor" as an alias for constructors.

Let's say we have this class, with a very long name

```csharp
public class MyClassWithVeryLongNameBecauseItsBetterForComprehension
{
}
```

Unfortunately, long class names have an undesired side-effect: they imply the creation of long-named constructors too.

```csharp
public class MyClassWithVeryLongNameBecauseItsBetterForComprehension
{
  public MyClassWithVeryLongNameBecauseItsBetterForComprehension()
  {
  }
}
```

This is acceptable for constructors without parameters, but it's less than perfect for constructors with (multiple) parameters.

```csharp
public class MyClassWithVeryLongNameBecauseItsBetterForComprehension
{
  public MyClassWithVeryLongNameBecauseItsBetterForComprehension(string parameter1, int parameter2, bool parameter3, char parameter4, float parameter5)
  {
  }
}
```

In this case, in we are forced to word-wrap the parameters into multiple lines in order to avoid horizontal scrolling, leading to a somehow less concise (and monitor unfriendly) representation of the constructor signature

```csharp
public class MyClassWithVeryLongNameBecauseItsBetterForComprehension
{
  public MyClassWithVeryLongNameBecauseItsBetterForComprehension(
      string parameter1, 
      int parameter2, 
      bool parameter3, 
      char parameter4, 
      float parameter5)
  {
  }
}
```

This could be easily avoided using a special, shorter keyword:
- "constructor" (like ECMAScript already does)
- "ctor"
- ".ctor" (like MSIL)

```csharp
public class MyClassWithVeryLongNameBecauseItsBetterForComprehension
{
  public constructor(string parameter1, int parameter2, bool parameter3, char parameter4, float parameter5)
  {
  }
}
```

As I mentioned, this is not someting new, it's already present in other languages, so it should be a familiar concept to a lot of coders.

What do you think ?

**Reply by HaloFour** [reply]
> the global evolution of the language lead to a "concise at any cost" approach

I don't believe that this is accurate.  Clarity is significantly more important than being terse.

Beyond that, I just don't think that using the `constructor` keyword here is more clear, or even more concise in most cases.  Assuming I ever typed out the full signature it would me more keystrokes in the majority of cases.  But more importantly I rarely have to type out the full signature due to the autocomplete features available in most IDEs, so the length of the class name is almost completely irrelevant.

**Reply by Rekkonnect** [reply]
> But more importantly I rarely have to type out the full signature due to the autocomplete features available in most IDEs, so the length of the class name is almost completely irrelevant.

As I've already pointed out, the main problem is when copy-pasting code across classes, usually in an inheritance chain. At least in VS, as of yet, there isn't a feature to quickly do this. Is relying on IDE capabilities a justification against having this improvement?

**Reply by NinjaCross** [reply]
> I don't believe that this is accurate. Clarity is significantly more important than being terse.

In general, I agree.
I just want to add that IMHO conciseness and clarity are strongly correlated, and often co-dependent.

Anyway...

> Beyond that, I just don't think that using the `constructor` keyword here is more clear, or even more concise in most cases. Assuming I ever typed out the full signature it would me more keystrokes in the majority of cases. But more importantly I rarely have to type out the full signature due to the autocomplete features available in most IDEs, so the length of the class name is almost completely irrelevant.

... well, you are not really disagreeing with me, we are talking about different things :)
You are talking about the implications during code **writing** .
I am more interested (and my proposal is more focused on) code **readability** and **formatting**.
As far as I can see, my point of view is not in contrast with yours: they are the two faces of the same medal.

**Reply by NinjaCross** [reply]
> > But more importantly I rarely have to type out the full signature due to the autocomplete features available in most IDEs, so the length of the class name is almost completely irrelevant.
> 
> As I've already pointed out, the main problem is when copy-pasting code across classes, usually in an inheritance chain. At least in VS, as of yet, there isn't a feature to quickly do this. Is relying on IDE capabilities a justification against having this improvement?

I agree with this, even if it wasn't one of my primary intents.
Code invariance in time is important.
The fewer parts of a class changes when I refactor it (i.e. rename it), the bettere it is.
Renaming a class using the `constructor` keyword would impact a single line of code into the whole CS file.
Without the `constructor` keyword, we need to change at least one line per constructor.
This is a very minor aspect, IMHO, but it's anyway something to consider.
A little advantage is better than no advantages at all :)

**Reply by foxesknow** [reply]
Rather than having a new keyword why not just use `new` ?
```

public class MyClassWithVeryLongNameBecauseItsBetterForComprehension
{
  public new(string parameter1, int parameter2, bool parameter3, char parameter4, float parameter5)
  {
  }
}
```

**Reply by Rekkonnect** [reply]
> Rather than having a new keyword why not just use `new` ?

Yes, that's what [I proposed](https://github.com/dotnet/csharplang/discussions/1928#discussioncomment-1329408)

**Reply by AJLProjects** [reply]
Don't use new becuase new makes sense for object instantiation, think in terms of English, create me a **new** cake. In this case, we aren't declarativley **creating** rather this is just the constructor, so a more appropriate name is init or ctor, or constructor(like js)

**Reply by Rekkonnect** [reply]
`private protected`, `void`, `using` (for variables) don't make that much sense in English either. That's not an argument in this case, for long as there is connection with the relevant feature, it works. `new` would mean that it's declaring the constructor that is only called after a `new` keyword. You won't ever call a constructor outside that context, and it's contained in the type that the constructor is for. You've already typed the name in the type's declaration, and you're declaring the method that will be invoked during an object creation, acting as its initialization.

Sure, constructors don't *actually* create the objects, they only *initialize* them. However, just like Java (`Object::new`), that's how you would break down an object creation:
```csharp
// an object creation
new Object();

// breaks down into this pseudocode
var memory = ReserveMemory();
memory.Initialize(Object.new);
```

**Reply by DavidArno** [reply]
F# uses `new` too for it's constructors and VB.NET uses `New`, so it's certainly set a precedent in other parts of the .NET ecosystem.

But for me, any time it's proposed to provide another way of doing the same thing, I have to ask myself "*If VS/R# suggested I change existing code from the old way to the new way, would I do it?*" For me, it's a clear "no" here. It's definitely a "meh" feature for me.

**Comment by jnm2** [comment]
@theunrepentantgeek My mistake, they are talking about declaration.

**Comment by ebfortin** [comment]
> @ebfortin Can you please clarify? Are you talking about use of the type name when _constructing_ an instance, or when _declaring_ the constructor as a part of the definition?
> 
> I thought it was the later, but the comment from @jnm2 has me questioning whether I read your proposal correctly.

When declaring.

**Comment by Austin-bryan** [comment]
@ebfortin Just type `ctor` then hit tab twice.

**Reply by DaniilPalii615** [reply]
Duplicated class name creates problems when reading code too. When I see in the code `public ClassName` and then see braces - I think "Oh, it is not class but constructor". And then while working with the class I should to keep in mind that `public ClassName() { ... }` it is `public constructor { ... }` in fact. Using `constructor` keyword would make work with code easier as properties instead of getters and setters.

**Comment by arekbal** [comment]
Answers here, seem to be very DRY. 
`ctor(), dtor()` ?
> Having two ways to do something, where the new way provides only a tiny amount of marginal benefit, would likely not be desirable.

Same goes with arrow syntax method definitions... but everyone agrees they are much more brief.
Java didn't want to change so Scala got created and then Kotlin. 
C# didn't want to change, so F# happened, C# suddenly started to change positively.
Pony, Rust, Midori, TypeScript happened, C# suddenly got interested in issues it didn't care about before.
I am not implying it was anybody's thought, just a general wave of programming world changes.
That thing (constructors repeating name of a class) is a low hanging fruit(almost no risk involved) on a  path to a more brief, easier to consume programming language.

**Comment by CyrusNajmabadi** [comment]
> Same goes with arrow syntax method definitions...

That was a case where there wasn't a "tiny amount of marginal benefit".  

> That thing (constructors repeating name of a class) is a low hanging fruit(almost no risk involved) on a path to a more brief, easier to consume programming language.

See: https://blogs.msdn.microsoft.com/ericgu/2004/01/12/minus-100-points/

Or, as Eric Lippert mentioned:

> And (2) just being a good feature is not enough. Features have to be so compelling that they are worth the enormous dollar costs of designing, implementing, testing, documenting and shipping the feature. They have to be worth the cost of complicating the language

> C# didn't want to change

This has never been the case.  

> Pony, Rust, Midori, TypeScript happened, C# suddenly got interested in issues it didn't care about before.

You are mistaking 'haven't been actively working on' with 'not caring'.  The C# LDM is interested in a wide variety of topics, and has been for decades.  But features *cost*.  It is literally zero sum.  If you take time and effort to work on a certain set of features, that time and effort is not available for others.  So what you see developed are the set of things that are felt to provide a great bang/buck ratio for the development ecosystem out there.

**Comment by arekbal** [comment]
It is always something or the other. You can't have all the cake right? 🍰 So it makes sense with silly and expensive features, or great but also very expensive features to lay them in priority queue and wait until they probably die off. I don't buy it. I don't see clear way features got selected. 
Seems very opinionated. Even _me_ going through some proposals, can see it.
Tell me, what enormous benefit (apart from brevity, which I very much value) syntax arrow definitions rewarded us. _Couple of characters less, doesn't reduce the amount of errors in code, actually increases them, because, from body you cannot tell if it returns or not. Leads to confusion between lambda function definition and static definition. Just use brackets dude. Don't forget about them._ 🐇
Don't get me started on IQueryable<> vs IEnumerable<>. ^^

**Comment by CyrusNajmabadi** [comment]
> I don't buy it. I don't see clear way features got selected.

It's not anyone's job to get you to buy anything :)  I'm simply explaining how language design is actually done.  And i'm doing that through my actual experience on the LDM team :)

> Seems very opinionated

Yes.  That's exactly right.  Language design is highly opinionated.  First, someone has to be opinionated enough to propose something.  Then an LDM member has to be opinionated enough to champion it.  Then, all hte opinions of the rest of the LDM are brought in during the design of the feature.  'Opinionated' is one hte best words i could come up with for the process :)

> Tell me, what enormous benefit (apart from brevity, which I very much value)

I don't see why the 'apart' part is relevant.  It's like saying: apart from the reasons you did this, what were the reasons you did this?

Brevity was the reason.  in general the existing statement form added two additional lines, and four extra tokens.  For the types of trivial, but common, methods that could use this feature this was a substantive overhead.  

> Don't get me started on IQueryable<> vs IEnumerable<>.

Ok.  Don't get started on that :)

**Comment by arekbal** [comment]
_In not so far future..._
Someone asks what was the point of bringing these shortcuts **ctor** and **dtor**(or similar), replacing **public** with **pub** and namespace-one-line-per-file-declaration instead of brackets:
>Brevity was the reason. in general the existing form added X additional lines, and Y extra tokens. For the types of trivial, but common, code that could use this feature it was a substantive overhead. Developers shouldn't have to repeat themselves as much.

**Comment by MV10** [comment]
#1335

**Comment by The-Futurist** [comment]
> It seems highly unlikely C# would change here. Having two ways to do something, where the new way provides only a tiny amount of marginal benefit, would likely not be desirable.

@CyrusNajmabadi 

Yet you exhibit support for [this ](https://github.com/dotnet/csharplang/issues/2098)which seems suffer from the same drawbacks: Another way of writing a switch that "provides only a tiny amount of marginal benefit" - or am I missing something?

**Comment by CyrusNajmabadi** [comment]
Yes.  As i mentioned in the other issue, the feature has already been championed.  Meaning the cost of it has already been assessed to be less than the gain overall to the language.  It's not the case of:

1. we have switch expressions, now someone wants to come along and add a new feature to it.

It's the case of:

1. we don't have switch expressions.  we're adding them.  what is the right design for them.

--

Here's another way of thinking about it.  Imagine any proposal has a fixed cost of 1000 LDM-bucks.  That's expensive and needs to present a lot of end value to the user for the LDM to pay that much.  Switch expressions are currently at having value of 10k LDM bucks.  So upping that cost to 10,005 to support consistency here is not really that different. 

However, having to go and spend the 1k LDM-bucks just to do a new way of doing constructors is just not there.  Maybe in the future if the LDM is doing a lot of constructor improvements, and the overall benefit is worth it, this could be rolled into that.  But paying so much to get a tiny marginal benefit is non-sensical.

**Comment by The-Futurist** [comment]
@CyrusNajmabadi 
> Yes. As i mentioned in the other issue, the feature has already been championed. Meaning the cost of it has already been assessed to be less than the gain overall to the language. It's not the case of:
> 

> 1. we have switch expressions, now someone wants to come along and add a new feature to it.
> 
> It's the case of:
> 
> 1. we don't have switch expressions.  we're adding them.  what is the right design for them.
> 
> --
> 
> Here's another way of thinking about it. Imagine any proposal has a fixed cost of 1000 LDM-bucks. That's expensive and needs to present a lot of end value to the user for the LDM to pay that much. Switch expressions are currently at having value of 10k LDM bucks. So upping that cost to 10,005 to support consistency here is not really that different.
> 

Of course I am in no position to even estimate cost so I can't comment on that, perhaps you could explain (if you have the time) the cost estimate for some of the proposals we make including this particular one? That can only help us going forward, to understand the true costs of some of the stuff we suggest here.

> However, having to go and spend the 1k LDM-bucks just to do a new way of doing constructors is just not there. Maybe in the future if the LDM is doing a lot of constructor improvements, and the overall benefit is worth it, this could be rolled into that. But paying so much to get a tiny marginal benefit is non-sensical.

I'm gonna take some convincing to see the utility in the trailing comma support. Your increasing the scope for unintended errors to go undiagnosed. Right now if someone edits code, copies, pastes etc and in error either removes the true end element or inserts something not intended to be the end element they'll get told off by the compiler.

With this change either scenario is regarded as valid and under the above user edit scenarios a potentially true error will go unreported and perhaps out into the production world. Any developer who claimed this was a real help to them in day to day work would also make me wonder about their skills.

So my vote (if that means anything, which I doubt!) is to deny this.

**Comment by CyrusNajmabadi** [comment]
> Of course I am in no position to even estimate cost so I can't comment on that, perhaps you could explain (if you have the time) the cost estimate for some of the proposals we make including this particular one? That can only help us going forward, to understand the true costs of some of the stuff we suggest here.

Every new feature has a high fixed cost (it's like passing a bill).  You have the meetings with everyone involved.  The impls, the testing, the context switching, the doc'ing, the shipping.  It's a huge fixed cost to even get a feature down this path *and* this is honestly close to zero-sum.  If the time is spent on htat, it's not spent on other things that deliver more value.

However, once you've decided to go down this path and you've paid the fixed costs, then changes/tweaks to something in flight is much less costly.  You already hve someone working on the design/grammar/impl/tests/etc. so saying: can you just tweak things to support X is much less costly.  Indeed, it basically just rolls into the normal cost that you've already accepted as being worth it to pay for this feature. 

Of course, if someone comes along and says "please tweak like so" and that tweaking doubles the cost, then it won't be considered the same as a tweak that basically takes 5 min of time.

**Comment by CyrusNajmabadi** [comment]
> I'm gonna take some convincing to see the utility in the trailing comma support. 
> So my vote (if that means anything, which I doubt!) is to deny this.

The change was approved.  The people who needed to be convinced were convinced :)

> Your increasing the scope for unintended errors to go undiagnosed. 

What unintended errors?

> Right now if someone edits code, copies, pastes etc and in error either removes the true end element or inserts something not intended to be the end element they'll get told off by the compiler.

I don't know what this means.  On hte other hand, we have ample experience that shows that supporting a trailing comma is not actually a problem in practice (i.e. all the other places in the language that allow this), and it makes for a definitely nicer experience across all editors.  It's not uncommon for people to keep needing to add elements to the end of a list that are very similar to the previous element.  Being able to just copy/paste that last element is a routine operation done by many devs across many IDEs/editors across many languages.  This routine op has a roadbump in it today that is just being removed.  Given the experience that shows this doesn't lead to problems, and given the nicer experience  this provides, it's simple to roll this work into the already approved feature which is coming down the pipe.

**Comment by The-Futurist** [comment]
> > Of course I am in no position to even estimate cost so I can't comment on that, perhaps you could explain (if you have the time) the cost estimate for some of the proposals we make including this particular one? That can only help us going forward, to understand the true costs of some of the stuff we suggest here.
> 
> Every new feature has a high fixed cost (it's like passing a bill). You have the meetings with everyone involved. The impls, the testing, the context switching, the doc'ing, the shipping. It's a huge fixed cost to even get a feature down this path _and_ this is honestly close to zero-sum. If the time is spent on htat, it's not spent on other things that deliver more value.
> 
> However, once you've decided to go down this path and you've paid the fixed costs, then changes/tweaks to something in flight is much less costly. You already hve someone working on the design/grammar/impl/tests/etc. so saying: can you just tweak things to support X is much less costly. Indeed, it basically just rolls into the normal cost that you've already accepted as being worth it to pay for this feature.
> 
> Of course, if someone comes along and says "please tweak like so" and that tweaking doubles the cost, then it won't be considered the same as a tweak that basically takes 5 min of time.

@CyrusNajmabadi 
So out of curiosity how do you express cost and compare costs? in person hours?

**Comment by The-Futurist** [comment]
I'll confine my remarks about the trailing comma to that thread, I apologize for derailing this one!

**Comment by CyrusNajmabadi** [comment]
> So out of curiosity how do you express cost and compare costs?

Historically, by the LDM members using their expertise and experience to weight/order things.  It's generally a continuous process, with certain points in product cycles being used to reasses/reorder things.

It's the same way i can read a proposal on a suggested IDE feature and generally give a very precise assessment of the cost to do that feature (including what sort of design tweaks could be considered, and how they would vary the bang/buck outcome of the work).  

Because nearly all of the LDM has actually worked with the language for years,and have been involved directly with the both the impl and the end-to-end process to start with a proposal and take it all the way to a released product, it's pretty easy for them to come to consensus here.

Note: not **everything** works out that way.  For example, NRT was something known to simply be so humongous and so impacting across teh language that it was known that the only way to even get a sense of it was to just 'do it'.  For something this large, this would then be a continual process of exploration/impl/assessment to see if the direction things were going, and the results that were being achieved would justify the costs involved.  But most everything else can be assessed apriori without needing this sort of approach.

Similarly, 'code generators' was felt to have absolutely staggering value to users.  However, early explorations into this space also revealed such a huge amount of unforseen work that this was tabled until more adequate funding and a more appropriate time could be found for this.

**Comment by AJLProjects** [comment]
Please can we re-review this? Clear violation of the DRY principle, and with probably millions of C# classes in existence, this is an impactful change. With regard to @CyrusNajmabadi last comment, how can you know if there is unseen work? This seems like an easy change to me.
UX is just as important as functionality when it comes to language design.

**Reply by canton7** [reply]
> Clear violation of the DRY principle

People continuously misinterpret DRY. DRY means ([from Wikipedia](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)):

> "Every **piece of knowledge** must have a single, unambiguous, authoritative representation within a system ... When the DRY principle is applied successfully, a modification of any single element of a system does not require a change in other logically unrelated elements

Typing the same thing multiple times is **not** necessarily a violation of DRY. DRY is about not duplicating bits of knowledge (commonly pieces of business logic) across your system.

Having the class name repeated in the class declaration and constructor declaration is not a violation of DRY.

**Reply by AJLProjects** [reply]
I just mean that you are forcing the user to repeat the same sybol. Last time I checked dry stood for **D**ont **R**epeat **Y**ourself.
Having the class name repeated in the class declaration and constructor is bad UX.

**Reply by HaloFour** [reply]
So then having the keyword `constructor` repeated for every constructor would also violate DRY.

**Reply by canton7** [reply]
Yes, the acronym "DRY" is misleading, and people often misinterpret it to mean "the same text should not appear more than once", which is very wrong.

Repeating a type name is not a violation of DRY, and saying that this is a "clear violation of the DRY principle" is incorrect.

> Having the class name repeated in the class declaration and constructor is bad UX.

I'm afraid you'll have to present some evidence for that: this language and others have managed perfectly fine for a long time.

**Reply by AJLProjects** [reply]
> So then having the keyword constructor repeated for every constructor would also violate DRY.

So by that logic every single time I write int, then I am repeating myself. 

The main thing is that you are repeating a self defined symbol, not a built in keyword. So then if I change the name of the class, then I will also have to change the name of the constructor, which is the same reason why we use var.

**Reply by AJLProjects** [reply]
> perfectly fine for a long time

This is just a carry over from C/C++ old days.

**Reply by canton7** [reply]
> So by that logic every single time I write int, then I am repeating myself.

Right, that's where you get if you extend your logic: I'm glad we agree that it doesn't make sense!

> So was slavery until it became incorrect.

That's a non-comparison: I'm going to ignore that.

> This is just a carry over from C++ old days.

Right, and so are lots of bits of C#. That doesn't automatically make them bad enough to change.

> So then if I change the name of the class, then I will also have to change the name of the constructor, which is the same reason why we use var.

Just use refactor-rename: that changes everything everywhere.

**Reply by AJLProjects** [reply]
refactor-rename is a compiler feature. The language itself should not have dependency on a compiler feature. 
Also that is the same argument against var.

var is to variables what 'constructor' keyword is to constructors

**Reply by canton7** [reply]
> refactor-rename is a compiler feature. The language itself should not have dependency on a compiler feature.

I think you mean an editor feature: the language very much depends on the compiler!

The compiler does provide language services for editors to use, and any-half decent editor has integration there, so it's kind of a moot point. If you're worried about renaming a type, and the number of places you need to change, surely you should be foccussing on having to name the type in field/properties and method parameters? Those are spread all over your code, whereas the constructor names are in the same file as the class declaration.

`var` is more about reducing redundancy on a single line when writing e.g. `Dictionary<string, List<string>> dict = new Dictionary<string, List<string>>()`. Some people see it as a way to change what type a method returns without having to go and fix up a lot of call sites, but there's controversy over whether this is a good thing, and a bit of a consensus saying that actually eliding type names everywhere leads to worse readability (which is why we're now getting editor features which show you the type of a `var`).

**Reply by AJLProjects** [reply]
Yes apologies, I meant editor. 
Yes you are correct var does deserve controversy becuase people are abusing it. Microsoft guidelines quickly changed to say only use var when the type is implied on same line. In this case constructor can only mean constructor, but var can mean any type.

**Reply by HaloFour** [reply]
Being DRY is not about never having the same content more than once in your code.  It's about not repeating the authoritative knowledge of that code.  The use of the editor to reduce manual repetition is explicitly encouraged in the same book that coined the term "DRY", including the use of source generators and refactoring.  It's not the language's job to eliminate the use of names of keywords from more than one location in your code, nor is it a principal of DRY to make it so that you never have to use the same name in more than one place.

**Reply by AJLProjects** [reply]
Ignore the dry thing, I'm ignorant.  But what business does the constructor name have with the name of the class? it can be contextually implied. Why are we creating an unnecessary dependency on the class name?

This is clearly subjective, since if one were objectivley correct, all lanaguages would be doing it one way.

Languages that require the user to repeat the class name in the constructor:
Scala, Java, C, C++, C#

Languages that infer the name/ use context:
Javascript uses constructor
Python uses __init__
Swift uses init
Ruby uses initialize

> It's not the language's job to eliminate the use of names of keywords from more than one location in your code

I agree. But did you mean ' eliminate the use of self defined types'? Since using constructor/init doesn't actually reduce numebr of symbols, it just changes the symbol dependency.

**Reply by canton7** [reply]
> This is clearly subjective

You're absolutely right. You need to be justifying why *your* subjective opinion is better than the subjective opinion of the language designers, and why it's worth going to the pain of moving the entire ecosystem over from one syntax to a different, equivalent syntax.

**Reply by CyrusNajmabadi** [reply]
Fwiw, if we were starting over again, I would push for this. However, given where we are now, I would not want this given the marginal benefit and the trillions of lines of code already out there using the current form.

**Reply by Rekkonnect** [reply]
This exact argument could be said for the simpler file namespace declaration that's coming in C# 10. Having gone this direction already, I cannot see why not implement this feature.

**Reply by CyrusNajmabadi** [reply]
Language design is not driven by 'why not', it's driven by 'why'.

**Reply by Rekkonnect** [reply]
The "why" has been already proposed in all the other comments, and that's the reason behind this proposal. The "why not" is what I argued against. What remains is quite a few "why"s, and no "why not"s, at least as I can see now.

**Reply by CyrusNajmabadi** [reply]
> The "why not" is what I argued against.

> Having gone this direction already, I cannot see why not implement this feature.

I'm having trouble reconciling these two statements.  In the first you say you're arguing against 'why not'.  But in the second, you do ask 'why not implement this feature?'... so you're arguing for 'why not' :)

**Reply by Rekkonnect** [reply]
Taking words a bit too literally, but I'll still explain.

I'm highly supportive of this feature, and even taken the time to write another comment regarding designing the feature. That being said, I've also taken the time to consider the "why yes" and "why not" arguments.

> Fwiw, if we were starting over again, I would push for this. However, given where we are now, I would not want this given the marginal benefit and the trillions of lines of code already out there using the current form.

You're the one that said you are hesitant to vote for this feature due to the fact that the language isn't fresh, but has been around for decades. And to that I say that the exact same argument **against** file-scoped namespace declarations could be used, but was not significant enough to prevent the feature from making it to the language. And for the record, I'm glad it did.

Even better, there will obviously be an available code fix to transform ctor declartions into their new form. There is literally no reason not to use it once it becomes available, as it's a full-fledged enhancement of an old design. Much like file-scoped namespace declarations, most people will opt-in by running a solution-wide code fix action and path forward. The old option is still there for the ones that don't like the new feature, or don't know about it, or don't care about it.

I hope this has been cleared up. "I don't see why not" in this case meant "we now have `whyNot.Count == 0`, so please reconsider your view because I support this feature, like lots of others". The design choices are always up to the LDT anyway, and it is not guaranteed if the syntax will make it the way it was initially proposed.

**Reply by CyrusNajmabadi** [reply]
> but was not significant enough to prevent the feature from making it to the language. 

Correct, in that case the reasons why were significant enough to warrant it.  So it was a 'why' decision.  We evaluated why we should do this, and found the benefits substantial enough.  I would want the same here.  

> Much like file-scoped namespace declarations, most people will opt-in by running a solution-wide code fix action and path forward

I have no reason to believe that this will be the case.  Indeed, we're not doing this in Roslyn itself (or any of the BCL libs afaict).

**Reply by Rekkonnect** [reply]
> Correct, in that case the reasons why were significant enough to warrant it. So it was a 'why' decision. We evaluated why we should do this, and found the benefits substantial enough. I would want the same here.

This now makes more sense, so the "why not" is not being removed from the list.

> I have no reason to believe that this will be the case.

I'm confident enough that many people do this for their projects if they really fancy the new features, once they choose to opt-in. And I personally do it, so I would greatly appreciate a diagnostic paired with a code fix for that purpose, it wouldn't be that hard to implement.

> Indeed, we're not doing this in Roslyn itself (or any of the BCL libs afaict).

I'm assuming you're referring to whether you're doing this refactoring in the code itself, not whether the code fix will be developed.

**Reply by CyrusNajmabadi** [reply]
> I'm assuming you're referring to whether you're doing this refactoring in the code itself,

Correct.

> not whether the code fix will be developed.

Correct.  The fixer is already available and shipped with the lang feature.

**Comment by deeprobin** [comment]
I think a syntax for static constructors or destructors wouldn't be bad either.

## Current
```cs
static MyClassName() {}

~MyClassName() {}
```

## My idea
```cs
static constructor { // I think these paratheses are in this case unnessecary because no static ctor or destructor can have parameters

}

destructor {
}
```

**Reply by Rekkonnect** [reply]
Building on top of this idea, since static constructors are not really "constructors" (they are conveniently called as such internally), but static initializer blocks, we could use something similar like:
```csharp
static init
{
}
```

**Reply by jnm2** [reply]
I'm not seeing a difference between the term "initializer block" and "constructor" for either the instance or static cases.

**Comment by ds5678** [comment]
Would this be considered an `Any Time` feature or a `Likely Never` feature? From reading the chat, it seems like the LDM is strongly opposed to devoting much time to this. However, would you accept a community contribution with a clean implementation?

**Reply by CyrusNajmabadi** [reply]
This is definitely not 'any time'.  It's also not 'likely never'.  It is in need of an LDM champion who would argue it through the LD process and get this accepted.  Note: i am not that person.  I'm personally against it (you can see my reasons in the thread).  But if someone else wanted it and convinced the LDM (including just getting it into the 'any time' bucket) then it could be done.

--

> However, would you accept a community contribution with a clean implementation?

It would not be accepted :)   Note, i'm trying to say that gently and not come across as dismissive of it.  While that would personally be really nice for you to do and would be appreciated (for an any-time feature), it's absolutely against the stated procedures and processes csharplang and Roslyn follow.  So i would not want you to go spend that effort and then end up sad if/when it was summarily rejected.
